\name{combine.geno.pheno}
\alias{combine.geno.pheno}
%- Also NEED an \alias for EACH other topic documented here.
\title{ 
	Combine the genotype and phenotype objects to create the base object for analysis by \code{\link{capeDO}}.
}
\description{
This function combines the phenotype object generated by \code{\link{read.pheno}} and the genotype object generated by \code{\\link{read.geno}} into a single object for analysis by the package
}
\usage{
combine.geno.pheno(geno, pheno)
}
%- maybe also usage for other objects documented here.
\arguments{
  \item{geno}{
	A genotype object generated by \code{\link{read.geno}}
}
  \item{pheno}{
	a phenotype object generated by \code{\link{read.pheno}}
	}
}
\value{
The genotype and phenotype objects are converted to a list object that is used as the main argument in most functions. The objects are inspected to ensure that individuals are present in both objects and are represented in the same order. 
This object is what is referred to as 'data.obj' in the argument lists of other functions. All data and analysis results will eventually be stored in this object. Upon creation the data.obj contains four elements:
\item{pheno}{A matrix containing the phenotype data for the population. Each phenotype is stored in a column, and individuals are stored in rows.}
\item{geno}{A three-dimensional array containing the genotype data for the population. The first dimension contains individuals, the second dimension contains alleles, and the the third dimension contains markers. Genotypes are stored probabilities between 0 and 1.}
\item{chromosome}{A vector containing the chromosome on which each marker is found.}
\item{marker.location}{A vector containing the chromosomal position of each marker.}
}
% Add one or more standard keywords, see file KEYWORDS in the
% R documentation directory.
\keyword{IO}

